### PR TITLE
[SPARK-48747][SQL] Add code point iterator to UTF8String

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
Add code point iteration capabilities to the UTF8String class.

### Why are the changes needed?
Need these changes to enable collation support for some expressions.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Testing new UTF8String features in `UTF8StringSuite`.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by allisonwang-db — COMMENTED** [review]
Hi @uros-db could you please create a JIRA ticket for this change?

**Comment by uros-db** [comment]
@allisonwang-db will do, thanks for the heads up

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -253,6 +255,16 @@ public int numBytes() {
    * Returns the number of code points in it.
    */
   public int numChars() {
+    if (numChars == 0) numChars = getNumChars();
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:258]: This looks like an unrelated change (to cache numChars). Can we do it with a separated PR? We should also add proper protection for multi-thread.

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:258]: agreed, removing this change from the current PR

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: opened a new PR for this: https://github.com/apache/spark/pull/47142

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -483,6 +583,46 @@ public byte getByte(int i) {
     return Platform.getByte(base, offset + i);
   }
 
+  /**
+   * Returns the code point at position `i`.
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:587]: we should be more clear about what position means here. It's the position of char, not byte

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by uros-db** [comment]
@cloud-fan ready to merge

**Comment by cloud-fan** [comment]
thanks, merging to master!
